### PR TITLE
Fix Issues 20530, 20537 - resolve trait isPackage/isModule and is(package/module) quirks

### PR DESCRIPTION
**Pull request by BorisCarvajal**
…ork with string mixins

#10739  continuation.

I made two solutions to the issue of the last comment but hit another bug while trying the long alternative so I decided to put that version on another pull to fix that bug.

The short version here is changing something I wondered in the past why it was there, the typeToExpression call is unnecessary and it gave me different results a few times on compiler experiments when using mixin("ident . ident ...") vs plain code, it seems the way that TypeIdentifier is converted to DotIdExp makes the code behave different on some cases.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by Geod24** [comment]
@thewilsonator : Please don't `auto-merge` so quickly. There's no way you could have reviewed it in 2 minutes. I don't see the point of just merging without review.  If we ever need this, we can just modify dlang-bot.
The previous solution had some issues, and was **also** prematurely merged before being reviewed. We might want to request a review from @MoonlightSentinel who actually did review and came up with good test cases last time.

**Comment by BorisCarvajal** [comment]
@Geod24: the previous solution was merged because the original bug report didn't have enough test cases, I didn't know about the test cases provided by @MoonlightSentinel.
Just saying.

**Comment by MoonlightSentinel** [comment]
> @Geod24: the previous solution was merged because the original bug report didn't have enough test cases,

A bug report isn't required to show of all possible inputs reproducing a certain bug. You can always add more test cases, especially if you are not completely sure about your fix (based on your initial comment https://github.com/dlang/dmd/pull/10739#issue-366669415)

> I didn't know about the test cases provided by @MoonlightSentinel.

These tests were taken from `compilable/test16002.d` by changing all simple identifiers to string mixins. Searching for existing tests in the test suite is usually a good start to find missing test cases.

File: test/compilable/test20530b.d
Comment (MoonlightSentinel) [inline test/compilable/test20530b.d]: I would prefer extending the original test case instead of scattering the tests across multiple files. This makes it easier to find missing test cases (and might reduce the runtime of the test suite)

File: src/dmd/traits.d
Comment (MoonlightSentinel) [inline src/dmd/traits.d]: `resolve` overrides the result of `isExpression` with `null` or another expression

File: test/compilable/test20530b.d
Comment (MoonlightSentinel) [inline test/compilable/test20530b.d]: Should probably have a test case with a full mixin and a string resulting from CTFE
(They should work regardless of these changes but just  to be sure)

File: src/dmd/traits.d
Comment (MoonlightSentinel) [reply src/dmd/traits.d]: Additionally this somehow breaks UFCS in speculative compilations:
```
void visit(alias what, T)(T t) {}

void main()
{
    int number;
    number.visit!(x => x + 1);
    static assert(__traits(compiles, visit!(x => x + 1)(number)));
    static assert(__traits(compiles, number.visit!(x => x + 1))); // fails
}
```

File: src/dmd/typesem.d
Comment (MoonlightSentinel) [inline src/dmd/typesem.d]: This is now unecessary when `t != null`

**Review by MoonlightSentinel — CHANGES_REQUESTED** [review]

File: src/dmd/traits.d
Comment (BorisCarvajal) [reply src/dmd/traits.d]: Yep, it seems that because the ambiguity in the context of trait-compiles and mixin, the code is parsed as a TypeIdentifier (which doesn't have UFCS) instead of DotIdExp, that's why the type is wrapped with typeToExpression. I've added test code in TypeIdentifier semantics to make it UFCS compatible as an experiment. But overall I'm discarding the approach of using this difference to fix the bug and I'm going to pull a new code.

File: test/compilable/test20530b.d
Comment (BorisCarvajal) [reply test/compilable/test20530b.d]: Ok, still gonna need a new file to test another bug that i'm attaching.

File: src/dmd/typesem.d
Comment (BorisCarvajal) [reply src/dmd/typesem.d]: yes my bad.

**Review by thewilsonator — APPROVED** [review]

**Comment by BorisCarvajal** [comment]
By the way, this patch also resolves the issue [11847](https://issues.dlang.org/show_bug.cgi?id=11847). Should I link the bug and upload the test case?
There is already a pull (#7218) open for that issue that seems to cover more about it. Maybe I should ask @MartinNowak if this patch has the chance to close that issue or a part of it at least.

**Comment by thewilsonator** [comment]
> Should I link the bug and upload the test case?

Yes please.

File: src/dmd/dmodule.d
Comment (WalterBright) [inline src/dmd/dmodule.d]: What is a "wrapping Package" ?

File: src/dmd/dmodule.d
Comment (WalterBright) [inline src/dmd/dmodule.d]: Use `isPackage()` instead of the cast. The cast is dangerous as `lookup` may return a different symbol.

File: src/dmd/dmodule.d
Comment (WalterBright) [inline src/dmd/dmodule.d]: Only the else case actually needs to be there. It'll simplify the logic.

File: src/dmd/dmodule.d
Comment (BorisCarvajal) [reply src/dmd/dmodule.d]: It's the Package that contains package.d and other submodules, the one inserted to the symbol table.
I took the name from a comment in this file:

> 2. Otherwise, 'package.d' wrapped by 'Package' is inserted to the internal tree in here.

File: src/dmd/dmodule.d
Comment (BorisCarvajal) [reply src/dmd/dmodule.d]: Done.

File: src/dmd/dmodule.d
Comment (BorisCarvajal) [reply src/dmd/dmodule.d]: Done.

File: src/dmd/dmodule.d
Comment (WalterBright) [reply src/dmd/dmodule.d]: I don't understand it there, either. Please add your explanation to the code comment.

File: src/dmd/dmodule.d
Comment (WalterBright) [reply src/dmd/dmodule.d]: much better

**Review by WalterBright — DISMISSED** [review]
Approved once definition of a "wrapper Package" is added.

File: src/dmd/dmodule.d
Comment (BorisCarvajal) [reply src/dmd/dmodule.d]: Do you mean adding this:
> It's the Package that contains package.d and other submodules, the one inserted to the symbol table.
Or a more detailed explanation?

Comment (BorisCarvajal) [reply src/dmd/dmodule.d]: Comment improved a little and I've added my full phrase to the block explaining the complete "package.d" insertion process .

**Comment by BorisCarvajal** [comment]
I made the requested changes.

**Comment by thewilsonator** [comment]
```
git clone -b v2.090.1 --depth=1 https://github.com/dlang/druntime .generated/druntime-2.090.1
Cloning into '.generated/druntime-2.090.1'...
warning: Could not find remote branch v2.090.1 to clone.
fatal: Remote branch v2.090.1 not found in upstream origin
```

Hmm, maybe try giving it a force push?

**Comment by BorisCarvajal** [comment]
> git clone -b v2.090.1 --depth=1 https://github.com/dlang/druntime .generated/druntime-2.090.1
> Cloning into '.generated/druntime-2.090.1'...
> warning: Could not find remote branch v2.090.1 to clone.
> fatal: Remote branch v2.090.1 not found in upstream origin

Sounds like a script bug, there is no `v2.090.1` tag yet, just `v2.090.1-beta.1`.

**Comment by thewilsonator** [comment]
Hmm, maybe try rebasing? Otherwise I'll have to defer to @CyberShadow or someone else about that.

**Comment by BorisCarvajal** [comment]
@thewilsonator, now I see the VERSION file was bumped but still no tag on dmd, druntime and phobos, I assume it's a matter of time.

**Comment by thewilsonator** [comment]
Try giving it a force push, then I guess.

**Pull request merged**
